kube-runtime/src/reflector/shared_store.rs

```diff
@@ -0,0 +1,401 @@
+            (None, lp2.clone().into()) => vec![Event::Restarted(expected_state2.clone())],
+        ));
+
+        let store1 = ss.all(lp1);
+        let store2 = ss.all(lp2);
+
+        ss.spawn().await;
+
+        assert_eq!(store1.cloned_state(), expected_state1, "Store 1");
+        assert_eq!(store2.cloned_state(), expected_state2, "Store 2");
+    }
+
+    #[tokio::test]
+    async fn returns_different_stores_for_same_list_params_with_different_scope() {
+        let lp = ListParams::default().labels("foo=bar");
+        let ns = "ns1";
+        let expected_state1 = vec![test_pod(1)];
+        let expected_state2 = vec![test_pod(2)];
+        let mut ss = TestSharedStore::<Pod>::new(hashmap!(
+            (None, lp.clone().into()) => vec![Event::Restarted(expected_state1.clone())],
+            (Some(ns.to_string()), lp.clone().into()) => vec![Event::Restarted(expected_state2.clone())],
+        ));
+
+        let cluster_store1 = ss.all(lp.clone());
+        let ns_store1 = ss.namespaced(ns, lp.clone());
+        let cluster_store2 = ss.all(lp);
+
+        ss.spawn().await;
+
+        assert_eq!(cluster_store1.cloned_state(), expected_state1, "ClusterStore 1");
+        assert_eq!(ns_store1.cloned_state(), expected_state2, "NS Store 1");
+        assert_eq!(cluster_store2.cloned_state(), expected_state1, "Cluster Store 2");
+    }
+
+    // TODO - Tests for the subscriptions
+
+    fn test_pod(postfix: usize) -> Pod {
+        Pod {
+            metadata: ObjectMeta {
+                name: Some(format!("test-pod-{}", postfix)),
+                namespace: Some("default".to_string()),
+                ..Default::default()
+            },
+            ..Default::default()
+        }
+    }
+
+    fn test_cm(postfix: usize) -> ConfigMap {
+        ConfigMap {
+            metadata: ObjectMeta {
+                name: Some(format!("test-pod-{}", postfix)),
+                namespace: Some("default".to_string()),
+                ..Default::default()
+            },
+            ..Default::default()
+        }
+    }
+
+    type TestSharedStore<K> = SharedStore<K, TestWatcherFactory<K>>;
+
+    impl<K> SharedStore<K, TestWatcherFactory<K>>
+    where
+        K: 'static + Resource + Clone + DeserializeOwned + Debug + Send + Sync,
+        K::DynamicType: Clone + Debug + Default + Eq + Hash + Unpin,
+    {
+        fn new(events: HashMap<(Option<String>, ListParams), Vec<Event<K>>>) -> Self {
+            Self {
+                watcher_provider: TestWatcherFactory {
+                    events: Mutex::new(events.into_iter().map(|(k, v)| (k, v.into())).collect()),
+                },
+                reflectors: HashMap::new(),
+            }
+        }
+
+        async fn spawn(self) {
+            tokio::spawn(async move {
+                self.run().for_each(|_| async {}).await;
+            });
+
+            tokio::task::yield_now().await;
+        }
```

Comment:
I think this call should be explicitly awaited in tests without spawning (and can probably be done in a one-liner) because there's nothing we actually wait for having mocked the provider - so there shouldn't be a need for a task.